### ISSUE TITLE
AWS Transfer Server: Attach VPC security groups at creation

### ISSUE THREAD

**Issue by ewbankkit** [hashicorp/terraform-provider-aws#15788]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

You can now attach multiple security groups to a server in a VPC.

### New or Affected Resource(s)

* [`aws_transfer_server`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/transfer_server)

### Potential Terraform Configuration

```hcl
resource "aws_transfer_server" "example" {
  endpoint_details {
    security_group_ids = [aws_security_group.example_1.id, aws_security_group.example_2.id]
  }
}
```

### References

[User Guide](https://docs.aws.amazon.com/transfer/latest/userguide/create-server-in-vpc.html).

Requires [AWS SDK v1.35.8](https://github.com/aws/aws-sdk-go/releases/tag/v1.35.8):

* #15500

Related:

* https://github.com/terraform-providers/terraform-provider-aws/pull/12599
* https://github.com/terraform-providers/terraform-provider-aws/pull/13737

[Note](https://docs.aws.amazon.com/transfer/latest/userguide/API_EndpointDetails.html#TransferFamily-Type-EndpointDetails-SecurityGroupIds):

> You can edit the SecurityGroupIds property in the UpdateServer API only if you are changing the EndpointType from PUBLIC or VPC_ENDPOINT to VPC. To change security groups associated with your server's VPC endpoint after creation, use the Amazon EC2 ModifyVpcEndpoint API.

**Comment by rodrigdav** [comment]
Any update on when this will be available?

**Comment by amadureira** [comment]
Hi @rodrigdav,

I have open a PR [17496](https://github.com/hashicorp/terraform-provider-aws/pull/17496).

**Comment by rodrigdav** [comment]
Thank you .  I will keep an eye out for it.

**Comment by amadureira** [comment]
Hi @rodrigdav and @zordrak,

I check my PR and I found some problems. I will  fix it. You could use my provider, see this configuration bellow:
```
source = "amadureira/aws"
version = "3.26.18"
```

**Comment by amadureira** [comment]
Hi  @ewbankkit , could you help-me?

I closed PR [17496](https://github.com/hashicorp/terraform-provider-aws/pull/17496) because the branch's name is wrong.  As mentioned on this [doc](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/pullrequest-submission-and-lifecycle.md#branch-prefixes), technical debt should start with **tb-** prefix.  I fixed others mistakes at this new PR [17539](https://github.com/hashicorp/terraform-provider-aws/pull/17539) but anyone saw it.  I don't know if I did something wrong.

Thanks for your help.

**Comment by redspeedy83** [comment]
Any news?

**Comment by GryshchenkoAndrii** [comment]
Really interested in this functionality. AWS cloudformation [supports](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-transfer-server-endpointdetails.html) it, would love support from terraform side as well.

**Comment by amadureira** [comment]
Hi Guys, 

I've applied the @ewbankkit recommendations.

I forgot to mark that is done. I do it now!

**Comment by rodrigdav** [comment]
I am starting to have some bandwidth.  I can test it out here in the next
few days.  Let me know how?

**Comment by sandeep3866** [comment]
any update on this, i don't see this feature is yet supported, looking for some response thanks

**Comment by ashish30jain** [comment]
Can we have this sooner released?  Its tweaky to update security group on endpoint without this.

**Comment by pspot2** [comment]
How should security groups be assigned in the meanwhile? After server creation (with the default SG), get to the VPC endpoint that was created and update the SGs there?

**Comment by MBarendregt** [comment]
Yep, that is how i do it.. i invoke a custom python script in tf to do that for me so it feels somewhat automated still :)

**Comment by armanbaghajyan** [comment]
When this will be available.

**Comment by jipumarino** [comment]
@MBarendregt how do you do that? When I try creating a VPC-hosted Transfer server Terraform just hangs. `TF_LOG=debug` shows these errors: 

```
Transfer/CreateServer failed, attempt 3/25, error InternalServiceError: Error calling CreateVpcEndpoint: An internal error has occurred
```

This is the config I'm trying it with: 
```
resource "aws_transfer_server" "test_vpc" {
  endpoint_type = "VPC"
  endpoint_details {
    subnet_ids = [...]
    vpc_id     = ...
  }
}
```

**Comment by quentin9696** [comment]
Hi, 

I'm also bocked by this feature. With this, impossible to create transfer server using shared VPC

In an aws account without vpc sharing, I create a vpc endpoint, then I use VPC_ENDPOINT and attached my transfer.server endpoint

**Comment by massettim** [comment]
Any news about it?

**Comment by ewbankkit** [comment]
This issue is on the [current quarter's roadmap](https://github.com/hashicorp/terraform-provider-aws/blob/main/ROADMAP.md#enhancements-to-existing-services).

**Comment by ewbankkit** [comment]
This has been addressed via #17539 (thanks to @amadureira for getting the solution started) and should be available in **v3.49.0** of the Terraform AWS Provider next week.
Note that the solution requires use of the EC2 `DescribeVpcEndpoints` and `ModifyVpcEndpoint` APIs for the Transfer Server `VPC` endpoint type, so IAM permissions should be modified appropriately - this is mentioned in the resource documentation.
